### PR TITLE
Add basic proxy auth

### PR DESCRIPTION
**Pull request by as0bu**
This allows the user to specify proxies during the chocolatey install.

### CODE REVIEW

File: templates/InstallChocolatey.ps1.erb [hunk 1]
```diff
@@ -38,9 +38,19 @@ param (
...
-  $downloader = new-object System.Net.WebClient
+  $downloader = New-Object System.Net.WebCLient
+<% if @proxy_server -%>
+  $proxy = New-Object System.Net.WebProxy("<% @proxy_server -%>:<% @proxy_port -%>", $true)
+  <%- if @proxy_username -%>
+  $passwd = ConvertTo-SecureString "<% @proxy_password -%>" -AsPlainText -Force; ## Website credentials
+  $proxy.Credentials = New-Object System.Management.Automation.PSCredential ("<% @proxy_username -%>", $passwd);
+  $downloader.UseDefaultCredentials = $true
+  <%- end -%>
+  $downloader.Proxy = $proxy
```
Comment (ferventcoder) [inline templates/InstallChocolatey.ps1.erb]: This looks very similar to what I just released at https://github.com/chocolatey/choco/blob/2fbecf671dd6a2bfc54ee2c6a97bb152649d340c/src/chocolatey.resources/helpers/functions/Get-WebFile.ps1#L40-L52

File: manifests/init.pp [hunk 2]
```diff
@@ -76,6 +80,25 @@
...
+    if ( ! $proxy_server) or ( ! $proxy_port) {
+      fail('Both proxy_port and proxy_server must be defined!')
+    }
+  }
+
+  if $proxy_username or $proxy_password {
+    if ( ! $proxy_password) or ( ! $proxy_username) or ( ! $proxy_server) {
+      fail('If proxy username or password is specified, all proxy settings must be defined!')
+    }
+  }
```
Comment (ferventcoder) [inline manifests/init.pp]: That's pretty nice. The specs will help drive out whether this is correct or not. :+1:

File: templates/InstallChocolatey.ps1.erb [hunk 3]
```diff
@@ -38,9 +38,19 @@ param (
   [string]$file
  )
   Write-Output "Downloading $url to $file"
-  $downloader = new-object System.Net.WebClient
+  $downloader = New-Object System.Net.WebCLient
+<% if @proxy_server -%>
+  $proxy = New-Object System.Net.WebProxy("<% @proxy_server -%>:<% @proxy_port -%>", $true)
+  <%- if @proxy_username -%>
```
Comment (ferventcoder) [inline templates/InstallChocolatey.ps1.erb]: This should probably validate proxy_password as well.

Comment (ferventcoder) [inline templates/InstallChocolatey.ps1.erb]: I see it was already done but would be nice to add. :)

File: templates/InstallChocolatey.ps1.erb (hunk 1 above)
Comment (as0bu) [inline templates/InstallChocolatey.ps1.erb]: I'm slightly confused about this. I thought that this was for strictly installing chocolatey itself via the puppet module. It looks like the link you sent was setting the proxy within chocolatey for packages to be installed via "choco install" correct? I did take this from https://github.com/chocolatey/choco/wiki/Proxy-Settings-for-Chocolatey and then just put it in the template file with some if statements.

Comment (ferventcoder) [inline templates/InstallChocolatey.ps1.erb]: @as0bu no worries, this was an FYI. It looks basically like what was implemented in chocolatey itself for installing packages and software.

Comment (as0bu) [inline templates/InstallChocolatey.ps1.erb]: @ferventcoder Oh ok I see!

**Comment by rismoney** [comment]
fwiw - i really like this PR.

**Comment by ferventcoder** [comment]
@rismoney agreed. :)

**Comment by ferventcoder** [comment]
I'm not sure what I did with my original comment, but @as0bu please add specs for this. This is great so far!

**Comment by as0bu** [comment]
Will do. Just give me a bit since I'm currently traveling back from a long distance client location. I hope to have that done Monday or Tuesday.

**Comment by mightydok** [comment]
@as0bu do you have time to finish this PR?

**Comment by ferventcoder** [comment]
@as0bu Ping - I know this was awhile ago. This would be a good addition to what you can pass, so if you can get those specs added and rebase this we can move forward!

**Comment by serialh0bbyist** [comment]
@ferventcoder I plan to do this during this week at PuppetConf.

**Comment by serialh0bbyist** [comment]
So after looking at this some more I think that commit e4c5f0945e41c79dab9c35cd5edf698987a13778 takes care of this. While the credentials aren't specified via Puppet, I think they're pulled from the operating system. @ferventcoder Could you confirm this?

**Comment by ferventcoder** [comment]
I think it may. We could always close this for now and if you see a need to pick this up in the future, we could do it then?

**Comment by serialh0bbyist** [comment]
Sounds like a plan!

**Comment by ferventcoder** [comment]
Closing for now.

**Comment by tardich** [comment]
Looks like a great puppet module. Unfortunately, despite all efforts I made, I can't make it to work. Updated the module to care about proxies (since I'm behind one), left all parameters to default (except for version which I set to 0.10.3), but I'm still getting this issue when trying to run:

`Error: Could not retrieve catalog from remote server: Error 400 on SERVER: undefined method 'ref' for nil:NilClass on node`

What do I do wrong?

**Comment by hbuckle** [comment]
I think this should be reopened as System.Net.WebClient will only pick up proxy settings from Internet Explorer and these are rather tricky to set automatically (requires calling WinInet in the Win32 API). A way to pass in an explicit proxy, either as a Puppet parameter or from an environment variable like the install script from chocolatey.org would be extremely useful.

**Comment by tardich** [comment]
At the same time, it would be interesting to be able to use the same level of complexity that are often in place (for example, at my place, we have complex proxy settings, all managed by the proxy.pac file. It is rather difficult for us to work with only one server / one port.

**Comment by ferventcoder** [comment]
@hbuckle @tardich explicit proxy would be great. However why don't we have you open an issue at https://tickets.puppet.com/browse/MODULES for the Puppet supported module - https://github.com/puppetlabs/chocolatey (so puppetlabs/chocolatey instead of chocolatey/chocolatey)? 

This issue is not an issue at all, it is a pull request. If you think this would be good to look at with chocolatey/chocolatey, then open an issue here to address.

**Comment by ferventcoder** [comment]
It does look like https://github.com/chocolatey/puppet-chocolatey/issues/142 was opened on this. You can comment/ subscribe to that issue.

**Pull request closed without merging**